contracts/cw20-streams/src/contract.rs

```diff
@@ -0,0 +1,446 @@
+    };
+    save_stream(deps, &stream)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "try_create_stream")
+        .add_attribute("owner", owner)
+        .add_attribute("recipient", recipient)
+        .add_attribute("amount", amount)
+        .add_attribute("start_time", start_time.to_string())
+        .add_attribute("end_time", end_time.to_string()))
+}
+
+pub fn execute_receive(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    wrapped: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if config.cw20_addr != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let msg: ReceiveMsg = from_binary(&wrapped.msg)?;
+    match msg {
+        ReceiveMsg::CreateStream {
+            recipient,
+            start_time,
+            end_time,
+        } => try_create_stream(
+            env,
+            deps,
+            wrapped.sender,
+            recipient,
+            wrapped.amount,
+            start_time,
+            end_time,
+        ),
+    }
+}
+
+pub fn try_withdraw(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    id: u64,
+) -> Result<Response, ContractError> {
+    let mut stream = STREAMS.load(deps.storage, id)?;
+    if stream.recipient != info.sender {
+        return Err(ContractError::NotStreamRecipient {});
+    }
+
+    if stream.claimed_amount >= stream.amount {
+        return Err(ContractError::StreamFullyClaimed {});
+    }
+
+    let block_time = env.block.time.seconds();
+    if stream.start_time >= block_time {
+        return Err(ContractError::StreamNotStarted {});
+    }
+
+    let unclaimed_amount = u128::from(block_time)
+        .checked_sub(stream.start_time.into())
+        .unwrap()
+        .checked_mul(stream.rate_per_second.u128())
+        .unwrap()
+        .checked_sub(stream.claimed_amount.u128())
+        .unwrap();
+
+    stream.claimed_amount = stream
+        .claimed_amount
+        .u128()
+        .checked_add(unclaimed_amount)
+        .unwrap()
+        .into();
+
+    STREAMS.save(deps.storage, id, &stream)?;
+
+    let config = CONFIG.load(deps.storage)?;
+    let cw20 = Cw20Contract(config.cw20_addr);
+    let msg = cw20.call(Cw20ExecuteMsg::Transfer {
+        recipient: stream.recipient.to_string(),
+        amount: unclaimed_amount.into(),
+    })?;
+
+    let res = Response::new()
+        .add_attribute("method", "try_withdraw")
+        .add_attribute("stream_id", id.to_string())
+        .add_attribute("amount", Uint128::from(unclaimed_amount))
+        .add_attribute("recipient", stream.recipient.to_string())
+        .add_message(msg);
+    Ok(res)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::GetConfig {} => to_binary(&query_config(deps)?),
+        QueryMsg::GetStream { id } => to_binary(&query_stream(deps, id)?),
```

Comment:
It would be nice to have a ListStreams{} query type, doing a range query with pagination. But that is an enhancement and can be done in another PR.

Reply:
Done

Reply:
Went ahead and implemented this.